src/protocol/prf_sharding/mod.rs

```diff
@@ -0,0 +1,704 @@
+
+    let mut output = vec![];
+
+    assert!(!input_rows.is_empty());
+    let first_row = &input_rows[0];
+    let mut prev_prf = first_row.prf_of_match_key;
+    let mut prev_row_inputs = initialize_new_device_attribution_variables(
+        Replicated::share_known_value(&binary_m_ctx, Gf2::ONE),
+        first_row,
+        num_breakdown_key_bits,
+        num_trigger_value_bits,
+        num_saturating_sum_bits,
+    );
+    let mut i: usize = 1;
+    let mut num_users_encountered = 0;
+    let mut row_for_user = 0;
+    while i < input_rows.len() {
+        let cur_row = &input_rows[i];
+        if prev_prf == cur_row.prf_of_match_key {
+            // Do some actual computation
+            let (inputs_required_for_next_row, capped_attribution_outputs) =
+                compute_row_with_previous(
+                    binary_m_ctx.clone(),
+                    RecordId(num_users_encountered),
+                    row_for_user,
+                    cur_row,
+                    &prev_row_inputs,
+                    num_breakdown_key_bits,
+                    num_trigger_value_bits,
+                    num_saturating_sum_bits,
+                )
+                .await?;
+            output.push(capped_attribution_outputs);
+            prev_row_inputs = inputs_required_for_next_row;
+
+            row_for_user += 1;
+        } else {
+            prev_prf = cur_row.prf_of_match_key;
+            prev_row_inputs = initialize_new_device_attribution_variables(
+                Replicated::share_known_value(&binary_m_ctx, Gf2::ONE),
+                cur_row,
+                num_breakdown_key_bits,
+                num_trigger_value_bits,
+                num_saturating_sum_bits,
+            );
+            row_for_user = 0;
+            num_users_encountered += 1;
+        }
+        i += 1;
+    }
+
+    Ok(output)
+}
+
+fn initialize_new_device_attribution_variables<BK, TV>(
+    share_of_one: Replicated<Gf2>,
+    input_row: &PrfShardedIpaInputRow<BK, TV>,
+    num_breakdown_key_bits: usize,
+    num_trigger_value_bits: usize,
+    num_saturating_sum_bits: usize,
+) -> InputsRequiredFromPrevRow
+where
+    BK: GaloisField,
+    TV: GaloisField,
+{
+    InputsRequiredFromPrevRow {
+        ever_encountered_a_source_event: share_of_one - &input_row.is_trigger_bit,
+        attributed_breakdown_key_bits: BitDecomposed::decompose(num_breakdown_key_bits, |i| {
+            Replicated::new(
+                Gf2::truncate_from(input_row.breakdown_key.left()[i]),
+                Gf2::truncate_from(input_row.breakdown_key.right()[i]),
+            )
+        }),
+        saturating_sum: BitDecomposed::new(vec![Replicated::ZERO; num_saturating_sum_bits]),
+        is_saturated: Replicated::ZERO,
+        // This is incorrect in the case that the CAP is less than the maximum value of "trigger value" for a single row
+        // Not a problem if you assume that's an invalid input
```

Comment:
Seems like a reasonable restriction, but probably worth an assertion or some other form of check?

Reply:
I've added:
```
assert!(num_saturating_sum_bits > num_trigger_value_bits);
```

So it should be impossible to encounter this error.